### PR TITLE
[WIP] add scrape_protocols for GlobalConfig and ScrapeConfig

### PR DESCRIPTION
**Pull request by dongjiang1989**
## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

ref: https://github.com/prometheus/prometheus/pull/12738 
## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
new option in GlobalConfig and ScrapeConfig: "scrape_protocols"
```

### CODE REVIEW

**Comment by slashpai** [comment]
This feature will be available in `2.48` Prometheus and it is not released yet. So we would need to wait for this till then

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 1]
```diff
@@ -219,6 +223,18 @@ type CommonPrometheusFields struct {
...
+	// The protocols to negotiate during a scrape. It tells clients what
+	// protocol are accepted by Prometheus and with what weight (most wanted is first).
+	// Supported values (case sensitive): PrometheusProto, OpenMetricsText0.0.1,
+	// OpenMetricsText1.0.0, PrometheusText0.0.4.
+	//
+	// If unset, Prometheus uses its default value: ["OpenMetricsText1.0.0","OpenMetricsText0.0.1","PrometheusText0.0.4"]
+	//
+	// It requires Prometheus >= v2.48.0.
+	//
+	// +optional
```
Comment (nicolastakashi) [inline pkg/apis/monitoring/v1/prometheus_types.go:235]: (nit): `+listType=atomic`

**Comment by dongjiang1989** [comment]
> This feature will be available in `2.48` Prometheus and it is not released yet. So we would need to wait for this till then

Thank you for your review. This PR depends on `2.48.x` Prometheus version.  Wait for this till  release

File: pkg/apis/monitoring/v1/prometheus_types.go (hunk 1 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1/prometheus_types.go:236]: Thanks for your review. Done.

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 2]
```diff
@@ -219,6 +223,19 @@ type CommonPrometheusFields struct {
 	// Number of seconds to wait until a scrape request times out.
 	ScrapeTimeout Duration `json:"scrapeTimeout,omitempty"`
 
+	// The protocols to negotiate during a scrape. It tells clients what
+	// protocol are accepted by Prometheus and with what weight (most wanted is first).
+	// Supported values (case sensitive): PrometheusProto, OpenMetricsText0.0.1,
+	// OpenMetricsText1.0.0, PrometheusText0.0.4.
+	//
+	// If unset, Prometheus uses its default value: ["OpenMetricsText1.0.0","OpenMetricsText0.0.1","PrometheusText0.0.4"]
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:231]: (nit) the default value might change, it's better to stay vague.
```suggestion
	// If unset, Prometheus uses its default value.
```

File: pkg/prometheus/promcfg.go [hunk 3]
```diff
@@ -1955,6 +1956,15 @@ func (cg *ConfigGenerator) appendScrapeIntervals(slice yaml.MapSlice) yaml.MapSl
 			Key: "scrape_timeout", Value: cpf.ScrapeTimeout,
 		})
 	}
+	return slice
+}
+
+func (cg *ConfigGenerator) appendScrapeProtocols(slice yaml.MapSlice) yaml.MapSlice {
+	cpf := cg.prom.GetCommonPrometheusFields()
+
+	if cpf.ScrapeProtocols != nil && len(cpf.ScrapeProtocols) > 0 {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1965]: (nit)
```suggestion
	if len(cpf.ScrapeProtocols) > 0 {
```

File: pkg/prometheus/promcfg.go [hunk 4]
```diff
@@ -2264,6 +2274,10 @@ func (cg *ConfigGenerator) generateScrapeConfig(
 		cfg = append(cfg, yaml.MapItem{Key: "scrape_timeout", Value: *sc.Spec.ScrapeTimeout})
 	}
 
+	if sc.Spec.ScrapeProtocols != nil && len(sc.Spec.ScrapeProtocols) > 0 {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2277]: (nit)
```suggestion
	if len(sc.Spec.ScrapeProtocols) > 0 {
```

**Review by simonpasquier — COMMENTED** [review]
Just a few nits, good work!

File: pkg/apis/monitoring/v1/prometheus_types.go (hunk 2 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1/prometheus_types.go]: Thank you for your opinion. Done

File: pkg/apis/monitoring/v1/prometheus_types.go (hunk 1 above)
Comment (ArthurSens) [reply pkg/apis/monitoring/v1/prometheus_types.go:236]: +listType was new to me, could you explain why choose atomic here?

[Reading the docs](https://book.kubebuilder.io/reference/markers/crd-processing) I'd say it is a set 🤔, but again, this is new to me
image

**Review by ArthurSens — APPROVED** [review]
Looking good! We just need to wait for 2.48 to be released :) 


Do you also want to add scrape protocols to ServiceMonitor and PodMonitors in a follow up PR?

**Comment by dongjiang1989** [comment]
> Looking good! We just need to wait for 2.48 to be released :)
> 
> Do you also want to add scrape protocols to ServiceMonitor and PodMonitors in a follow up PR?

@ArthurSens  Yes. I would love to add scrape protocols to ServiceMonitor and PodMonitors in a follow up PR.

File: pkg/apis/monitoring/v1/prometheus_types.go (hunk 1 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1/prometheus_types.go:236]: https://github.com/kubernetes/kube-openapi/blob/master/pkg/idl/doc.go  `atomic` as default value for enum

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 5]
```diff
@@ -219,6 +223,19 @@ type CommonPrometheusFields struct {
...
+	// The protocols to negotiate during a scrape. It tells clients what
+	// protocol are accepted by Prometheus and with what weight (most wanted is first).
+	// Supported values (case sensitive): PrometheusProto, OpenMetricsText0.0.1,
+	// OpenMetricsText1.0.0, PrometheusText0.0.4.
+	//
+	// If unset, Prometheus uses its default value.
+	//
+	// It requires Prometheus >= v2.48.0.
+	//
+	// +listType=atomic
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:235]: Values can't be deduplicated.
```suggestion
	// +listType=set
```

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 6]
```diff
@@ -114,6 +114,18 @@ type ScrapeConfigSpec struct {
...
+	// The protocols to negotiate during a scrape. It tells clients what
+	// protocol are accepted by Prometheus and with what weight (most wanted is first).
+	// Supported values (case sensitive): PrometheusProto, OpenMetricsText0.0.1,
+	// OpenMetricsText1.0.0, PrometheusText0.0.4.
+	//
+	// If unset, Prometheus uses its default value: ["OpenMetricsText1.0.0","OpenMetricsText0.0.1","PrometheusText0.0.4"]
+	//
+	// It requires Prometheus >= v2.48.0.
+	//
+	// +listType=atomic
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:126]:
```suggestion
	// +listType=set
```

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 7]
```diff
@@ -30,6 +30,10 @@ const (
 	PrometheusKindKey = "prometheus"
 )
 
+// ScrapeProtocol represents supported protocol for scraping metrics.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:33]: Let's list the enum values explicitly in the docs:
```suggestion
// ScrapeProtocol represents a protocol used by Prometheus for scraping metrics.
// Supported values are:
// * `OpenMetricsText0.0.1`
// * `OpenMetricsText1.0.0`
// * `PrometheusProto`
// * `PrometheusText0.0.4`
````

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 8]
```diff
@@ -219,6 +223,19 @@ type CommonPrometheusFields struct {
 	// Number of seconds to wait until a scrape request times out.
 	ScrapeTimeout Duration `json:"scrapeTimeout,omitempty"`
 
+	// The protocols to negotiate during a scrape. It tells clients what
+	// protocol are accepted by Prometheus and with what weight (most wanted is first).
+	// Supported values (case sensitive): PrometheusProto, OpenMetricsText0.0.1,
+	// OpenMetricsText1.0.0, PrometheusText0.0.4.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:229]:
```suggestion
```

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 9]
```diff
@@ -114,6 +114,18 @@ type ScrapeConfigSpec struct {
 	// ScrapeTimeout is the number of seconds to wait until a scrape request times out.
 	// +optional
 	ScrapeTimeout *v1.Duration `json:"scrapeTimeout,omitempty"`
+	// The protocols to negotiate during a scrape. It tells clients what
+	// protocol are accepted by Prometheus and with what weight (most wanted is first).
+	// Supported values (case sensitive): PrometheusProto, OpenMetricsText0.0.1,
+	// OpenMetricsText1.0.0, PrometheusText0.0.4.
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:120]:
```suggestion
```

**Comment by ArthurSens** [comment]
Also, it looks like `scrape_protocols` haven't made it into 2.48 :(

We'll have to wait another release

File: pkg/apis/monitoring/v1/prometheus_types.go (hunk 7 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1/prometheus_types.go:34]: Thanks for your review. 
Done

File: pkg/prometheus/promcfg.go [hunk 10]
```diff
@@ -1988,6 +1989,15 @@ func (cg *ConfigGenerator) appendScrapeIntervals(slice yaml.MapSlice) yaml.MapSl
...
 		})
 	}
+	return slice
+}
+
+func (cg *ConfigGenerator) appendScrapeProtocols(slice yaml.MapSlice) yaml.MapSlice {
+	cpf := cg.prom.GetCommonPrometheusFields()
+
+	if len(cpf.ScrapeProtocols) > 0 {
+		slice = cg.WithMinimumVersion("2.48.0").AppendMapItem(slice, "scrape_protocols", cpf.ScrapeProtocols)
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1999]:
```suggestion
		slice = cg.WithMinimumVersion("2.49.0").AppendMapItem(slice, "scrape_protocols", cpf.ScrapeProtocols)
```

File: pkg/prometheus/promcfg.go [hunk 11]
```diff
@@ -2297,6 +2307,10 @@ func (cg *ConfigGenerator) generateScrapeConfig(
 		cfg = append(cfg, yaml.MapItem{Key: "scrape_timeout", Value: *sc.Spec.ScrapeTimeout})
 	}
 
+	if len(sc.Spec.ScrapeProtocols) > 0 {
+		cfg = cg.WithMinimumVersion("2.48.0").AppendMapItem(cfg, "scrape_protocols", sc.Spec.ScrapeProtocols)
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2311]:
```suggestion
		cfg = cg.WithMinimumVersion("2.49.0").AppendMapItem(cfg, "scrape_protocols", sc.Spec.ScrapeProtocols)
```

**Review by simonpasquier — CHANGES_REQUESTED** [review]

File: pkg/prometheus/promcfg.go (hunk 11 above)
Comment (dongjiang1989) [reply pkg/prometheus/promcfg.go:2311]: Got it! I'm waiting for the final release version.

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 12]
```diff
@@ -230,6 +239,17 @@ type CommonPrometheusFields struct {
 	// Number of seconds to wait until a scrape request times out.
 	ScrapeTimeout Duration `json:"scrapeTimeout,omitempty"`
 
+	// The protocols to negotiate during a scrape. It tells clients what
+	// protocol are accepted by Prometheus and with what weight (most wanted is first).
+	//
+	// If unset, Prometheus uses its default value.
+	//
+	// It requires Prometheus >= v2.48.0.
```
Comment (ArthurSens) [inline pkg/apis/monitoring/v1/prometheus_types.go:247]:
```suggestion
	// It requires Prometheus >= v2.49.0.
```

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 13]
```diff
@@ -132,6 +132,16 @@ type ScrapeConfigSpec struct {
 	// ScrapeTimeout is the number of seconds to wait until a scrape request times out.
 	// +optional
 	ScrapeTimeout *v1.Duration `json:"scrapeTimeout,omitempty"`
+	// The protocols to negotiate during a scrape. It tells clients what
+	// protocol are accepted by Prometheus and with what weight (most wanted is first).
+	//
+	// If unset, Prometheus uses its default value: ["OpenMetricsText1.0.0","OpenMetricsText0.0.1","PrometheusText0.0.4"]
+	//
+	// It requires Prometheus >= v2.48.0.
```
Comment (ArthurSens) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:140]:
```suggestion
	// It requires Prometheus >= v2.49.0.
```

**Review by ArthurSens — COMMENTED** [review]
2.49 was release, we're ready to go here 🚀

**Comment by ArthurSens** [comment]
Ah, just noticed that you've opened #6235 , should we close this one then?

**Comment by ArthurSens** [comment]
Closing this one since #6235 was merged! Please let me know if I'm mistaken and this should stay open :)

**Pull request closed without merging**